assets/model_monitoring/components/src/model_data_collector_preprocessor/span_tree_utils.py

```diff
@@ -0,0 +1,168 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+"""Span Tree internal utilities and classes for building a Span Tree from preprocessed span logs."""
+
+
+import bisect
+from datetime import datetime
+import json
+
+from typing import Iterator, List
+from pyspark.sql import Row
+
+
+class SpanTreeNode:
+    """Spantree node class."""
+
+    def __init__(self, span_row: Row) -> None:
+        """Represent a singular node in a span tree."""
+        self._span_row = span_row
+        self._children = []
+
+    @property
+    def span_id(self) -> str:
+        """Get the span id."""
+        return self._span_row.span_id
+
+    @property
+    def parent_id(self) -> str:
+        """Get the span's parent id."""
+        return self._span_row.parent_id
+
+    @property
+    def children(self) -> List["SpanTreeNode"]:
+        """Get the span's children as list."""
+        return self._children
+
+    @children.setter
+    def children(self, value: list) -> None:
+        """Set the span's children."""
+        self._children = value
+
+    @property
+    def span_row(self) -> Row:
+        """Get the span's internal row."""
+        return self._span_row
+
+    @classmethod
+    def create_node_from_json_str(cls, json_str: str) -> "SpanTreeNode":
+        """Parse json string to get a single SpanTree node."""
+        node_dict: dict = json.loads(json_str)
+
+        node_dict['start_time'] = datetime.fromisoformat(node_dict['start_time'])
+        node_dict['end_time'] = datetime.fromisoformat(node_dict['end_time'])
+        children = node_dict.pop('children', [])
+        new_row = Row(**node_dict)
+
+        obj = cls.__new__(cls)
+        super(SpanTreeNode, obj).__init__()
+        obj._span_row = new_row
+        obj._children = children
+        return obj
+
+    def insert_child(self, span: "SpanTreeNode") -> None:
+        """Insert a child span in ascending time order due to __lt__()."""
+        bisect.insort(self._children, span)
+
+    def show(self, indent: int = 0) -> None:
+        """Print the current span in a formatted syntax to stdout."""
+        print(f"{' '*indent}[{self.span_row.span_id}({self.span_row.start_time}, {self.span_row.end_time})]")
+        for c in self.children:
+            c.show(indent + 4)
+
+    def to_dict(self) -> dict:
+        """Get dictionary representation of SpanTreeNode."""
+        span_row_dict = self.span_row.asDict()
+        span_row_dict['children'] = self.children
+        return span_row_dict
+
+    def __iter__(self) -> Iterator["SpanTreeNode"]:
+        """Iterate over current span and child spans."""
+        for child_span in self._children:
```

Comment:
we should use `self._children or []` in case _children is None

Reply:
will add in v2